[FEAT] 소셜 로그인 회원 통합 계정 기능

jye326 opened the pull request:
## Issue Number
closed #1050 

## 문제

ID·PW 방식으로 회원 가입한 이력이 있는 유저는 전화번호 유니크 정책으로 인해 소셜 로그인·회원가입이 불가능
- 이미 존재하는 전화번호로 소셜 로그인 계정 생성 시도 시, 400 응답을 받는다.

<img width="1500" height="600" alt="Image" src="https://github.com/user-attachments/assets/7c7c785d-2372-4b62-9296-27799a751c56" />

## 액션

- 이미 존재하는 회원의 전화번호로 인증을 완료한 후 Oauth 회원가입 요청을 보낼 시, 해당 회원의 기존 Member 객체를 참조하는 oauthMember 객체를 생성하여 소셜 회원가입한다.

## 해결
- 기존에 ID·PW 방식 회원가입을 한 유저도 같은 전화번호로 소셜 로그인을 이용할 수 있다.

## 추가 고민
- 반대로 Oauth로 회원가입을 한 유저가 ID·PW 방식으로 회원 가입을 할 수는 없는 문제가 남아있다.
- 전화번호 중복 시 Oauth 회원을 등록하는 것은 가능하지만 전화번호 중복 시 일반 회원 정보를 재등록 할 수는 없기 때문이다.
- 이는 추가 UI와 함께 정책적으로 고민해야 할 것 같다.

## MockitoSpyBean
- 테스트를 위해 JwtProvider 객체를 MockitoSpyBean으로 등록하였다. MockitoSpyBean은 기본적으로 실제 Bean을 띄우지만
- willReturn(), doReturn() 구문으로 Mocking 메서드를 호출한 테스트에서는 모킹 빈을 호출한다.
- 따라서 oauth 회원가입 테스트에서만 JwtProvider 객체를 모킹할 수 있다.

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?

jye326 committed:
fix: 카카오 로그인 및 회원가입 로직 리팩토링

- `existsByPhone_Number` 메서드명을 `existsByPhoneNumber`로 변경
- 카카오 로그인 로직 분리 (`allowOauthSignIn`, `allowOauthSignUp`)
- `findByPhoneNumber` 메서드 추가
- `oauthSignUp` 컨트롤러 매서드 가독성 향상 (파라미터 정렬)

jye326 committed:
fix: 카카오 회원가입 로직 수정

- 기존 회원 존재 시 MemberOauth만 생성하여 저장하도록 수정
- 신규 회원 생성 시 별도 메서드(getRandomIdPwMember)로 분리
- 불필요한 로직 및 중복 제거으로 코드 간소화

jye326 committed:
fix: MemberRepository 메서드명 수정

- `existsByPhoneNumber`를 `existsByPhone_Number`로 수정
- `findByPhoneNumber`를 `findByPhone_Number`로 수정

jye326 committed:
feat: oauth 회원가입 테스트 추가

- oauth 회원가입 가능 테스트 추가
- 기존 회원의 oauth 회원가입 가능 테스트 추가
- JwtProvider mocking 적용

jye326 committed:
test: JwtProvider SpyBean 적용 및 Oauth 회원가입 테스트 수정

- IntegrationTestSupport에서 JwtProvider를 MockitoSpyBean으로 변경
- AuthServiceTest에서 willReturn 메서드를 사용한 jwtProvider mocking 수정
- 기존 mockJwtProvider 대신 jwtProvider를 활용하도록 테스트 코드 변경

kimdevv commented on backend/fittoring/src/main/java/fittoring/application/auth/service/AuthService.java:
```diff
@@ -118,21 +118,25 @@ public void logout(Long memberId) {
+        return allowOauthSignUp(kakaoId);
+    }
 
+    private LoginInfoDto allowOauthSignIn(Optional<MemberOauth> memberOauth) {
```
p1: 혼동을 방지하기 위해 SignIn 말고 LogIn으로 통일하는 게 좋을 것 같습니다.

jye326 replied:
통일성 있게 signUp/signIn으로 맞추고 싶었는데 기존 코드가 이미 register/login 으로 작성되어 있으니 맞추겠습니다!

kimdevv commented on backend/fittoring/src/test/java/fittoring/application/auth/service/AuthServiceTest.java:
```diff
@@ -211,4 +212,45 @@ void logout2() {
+        // given
+        String phoneNumber = "010-1234-5678";
+        OauthSignUpRequest request = new OauthSignUpRequest("이름", "MALE", phoneNumber);
+        willReturn(1L).given(jwtProvider).getSubjectFromPayloadBy(any());
```
p3: `@MockitoSpyBean`을 적용해서 `jwtProvider`에서는 `getSubjectFromPayloadBy()` 메서드만 모킹되는 거군요.

jye326 replied:
맞습니다. `SpyBean`은 부분 메서드만 모킹하고 나머지는 실제 `Bean`을 사용합니다.

kimdevv reviewed (commented):
고생하셨습니다~ 간단한 커멘트 하나 남깁니다. 확인 부탁드려요!

jye326 committed:
refactor: AuthService 메서드 명칭 수정으로 일관성 향상

- allowOauthSignIn -> allowOauthLogin, allowOauthSignUp -> allowOauthRegistration으로 메서드 이름 변경
- 코드 명확성과 가독성 개선

dye0p commented on backend/fittoring/src/main/java/fittoring/application/auth/service/AuthService.java:
```diff
@@ -141,6 +145,22 @@ public LoginInfoDto kakaoLogin(String code) {
+        memberRepository.save(member);
+        MemberOauth memberOauth = new MemberOauth(member, AuthProvider.KAKAO, oauthId);
+        memberOauthRepository.save(memberOauth);
+        return memberOauth;
```
p2: 엔티티를 컨트롤러로 반환시키도록 구현한 이유가 궁금합니다~

jye326 replied:
계층 분리를 위해 Dto를 반환하도록 리팩터링하였습니다!

dye0p commented on backend/fittoring/src/main/java/fittoring/application/auth/service/AuthService.java:
```diff
@@ -151,11 +171,7 @@ public MemberOauth registerOauthMember(OauthSignUpRequest request, String oauthS
-        MemberOauth memberOauth = new MemberOauth(member, AuthProvider.KAKAO, oauthId);
-        memberOauthRepository.save(memberOauth);
-        return memberOauth;
+        return member;
```
```suggestion
    private Member getRandomIdPwMember(OauthSignUpRequest request) {
        String randomLoginId = RandomStringUtils.randomAlphanumeric(20);
        validateDuplicateLoginId(randomLoginId);
        String randomPw = RandomStringUtils.randomAlphanumeric(20);
        return new Member(
                randomLoginId,
                request.gender(),
                request.name(),
                new Phone(request.phone()),
                Password.from(randomPw)
        );
    }

```
`validateDuplicateLoginId`의 순서를 바꾸면 코드가 조금 간결해질 것 같아요!

jye326 replied:
반영했습니다~!

dye0p commented on backend/fittoring/src/test/java/fittoring/IntegrationTestSupport.java:
```diff
@@ -24,6 +26,9 @@ public abstract class IntegrationTestSupport {
     @MockitoBean
     protected MentoringPaginationHelper mph;
 
+    @MockitoSpyBean
+    protected JwtProvider jwtProvider;
+
```
p3: `@MockitoSpyBean`을 사용한 이유가 궁금하니다! `@MockBean`도 사용할 수 있을 것 같아서요~

jye326 replied:
다른 메서드는 모킹하지 않고 실제 객체를 사용하도록 `@MockitoSpyBean`을 사용했습니다!

PR에 추가된 테스트는 `getSubjectFromPayloadBy()` 만 사용해서 `@MockBean`을 사용해도 실행이 가능하지만 기존에 존재하던 테스트들이 JwtProvider의 다른 메서드들을 호출하거든요! 모든 메서드를 각 테스트별로 일일이 모킹하지 않고 필요한 메서드만 모킹해서 사용하기 위함입니다!

dye0p reviewed (commented):
수고하셨습니다! 몇가지 질문 남겼으니 확인해주세요~

jye326 committed:
refactor: OAuth 회원가입 로직 리팩토링 및 RegisterOAuthDto 도입

- registerOauthMember 메서드에서 기존 MemberOauth 반환을 RegisterOAuthDto 반환으로 변경
- RegisterOAuthDto 클래스 추가로 반환 데이터 구조화
- AuthService, AuthController, AuthServiceTest에 관련 코드 수정 및 리팩토링
- 코드 가독성 및 데이터 구조 명확성 향상

jye326 committed:
Merge branch 'develop' into feature/1050-auth-integration

jye326 committed:
refactor: OauthSignUpRequest gender 필드 enum 타입으로 변경

- Gender.STRING에서 Gender.ENUM으로 변경하여 코드 명확성 향상
- AuthServiceTest 관련 테스트 코드 수정 및 리팩토링

dye0p reviewed (approved):
특이사항은 안보이는 것 같아요~ 승인하겠습니다!

kimdevv reviewed (approved):
고생하셨습니다. 저도 이만 승인할게요~
